### PR TITLE
uboot-kirkwood: fix malformed boot configuration

### PR DESCRIPTION
**Pull request by bobafetthotmail**
With current boot default configuration the uboot will
fail to start the OpenWrt firmware with the following error:

```
unexpected character 'b' at the end of partition
Error initializing mtdparts!
incorrect device type in ubi
Partition ubi not found!
Error, no UBI device/partition selected!
Wrong Image Format for bootm command
Error occured, error code = 112
```

If the uboot configuration is examined with printenv
I can see that mdtparts line (on a nsa310) is wrong:

`mtdparts=mtdparts=orion_nand:0x0c0000(uboot),0x80000(uboot_env),0x7ec0000(ubi)bootargs_root=`
The "bootargs_root=" that was appended to it should not be there.

Fix the issue by adding a \0 line terminator to all affected lines.

This issue was detected and the fix tested on a nsa310, nsa325 and
a pogoplug v4, but it's not hardware-specific, so apply the same fix
to other devices as well.

Note that the issue is with the uboot's integrated boot configuration,
which is not used unless the uboot configuration in flash is unavailable
(erased or corrupted), which happens only on first time installation,
or if the user deletes the uboot configuration when upgrading uboot.
People just upgrading from an older uboot without erasing their previous
uboot configuration stored in flash would not have noticed this issue.

Signed-off-by: Alberto Bursi <alberto.bursi@outlook.it>

This was reported (and there are longer logs) in this forum thread https://forum.lede-project.org/t/installing-lede-on-nsa325/3611/44

@p-wassi might be interested in this.

@blogic or @mkresin can this also be ported to 18.06 branch too? uboot generated without this fix are unable to boot OpenWrt if someone is installing it on a new device.

### CODE REVIEW

**Comment by bobafetthotmail** [comment]
The guy that reported the issue confirmed it is good also on his device (a nsa310)
https://forum.lede-project.org/t/installing-lede-on-nsa325/3611/53

File: package/boot/uboot-kirkwood/patches/007-nsa310-uboot-generic.patch [hunk 1]
```diff
@@ -28,20 +28,20 @@ NOTE: this patch is ready for upstream, LEDE-specific parts are in
 @@ -56,6 +56,9 @@ config TARGET_GOFLEXHOME
  config TARGET_NAS220
  	bool "BlackArmor NAS220"
- 
+
```
Comment (blogic) [inline package/boot/uboot-kirkwood/patches/007-nsa310-uboot-generic.patch]: drop all these unrelated changes please and put them into a separate patch

**Comment by bobafetthotmail** [comment]
done

**Comment by bobafetthotmail** [comment]
ping @blogic

**Comment by mkresin** [comment]
Hey Alberto,

comparing your approach with [what can be found upstream](http://git.denx.de/?p=u-boot.git&a=search&h=HEAD&st=grep&s=CONFIG_MTDPARTS), it looks like you are holding it wrong ;-).

Upstream adds the "\0" to the CONFIG_EXTRA_ENV_SETTINGS define. Considering the following chunk in 008-nsa325-uboot-generic.patch

```diff
+#define CONFIG_EXTRA_ENV_SETTINGS \
+	"console=console=ttyS0,115200\0"	\
+	"mtdids=nand0=orion_nand\0"		\
+	"mtdparts="CONFIG_MTDPARTS_DEFAULT	\
+	"bootargs_root=\0"
```

it looks indeed looks like the more suitable place to add the zero terminator.

Hopefully I haven't missed something and what I wrote makes sense.

**Comment by mkresin** [comment]
I really don't want to bikesheed the PR, but the commit message of "uboot-kirkwood: fix malformed boot configuration" is about adding the missing "\0", while it changes the mtdparts for some boards completely.

Right now it looks like these changes are slipped in and are not intentional. The best would be to move them into an extra commit, with a proper explanation why the changed mtdparts are required.

**Comment by bobafetthotmail** [comment]
> Upstream adds the "\0" to the CONFIG_EXTRA_ENV_SETTINGS define.

That's ok for me, will respin this PR in a few hours.

> while it changes the mtdparts for some boards completely.

No, this PR is "a patch modifying patches", and github interface is too basic and gets confused.

For example, this is one of the patches that according to Github I altered the mtdparts of:

in its original form in master
https://github.com/openwrt/openwrt/blob/master/package/boot/uboot-kirkwood/patches/110-dockstar.patch

in my PR
https://github.com/bobafetthotmail/openwrt/blob/709c6c5ac6959768210a6c91d4edf725c0eae259/package/boot/uboot-kirkwood/patches/110-dockstar.patch

The only difference is the addition of \0 where I said I put it.

**Comment by bobafetthotmail** [comment]
Ok, updated. Now it does the same as upstream. @mkresin 

Github is still very confused by this "patch of a patch" so it still shows that I changed stuff I did not.

As said in the PR message, this should be applied also in 18.06 branch, since it would prevent booting when installing openwrt for the first time in kirkwood devices affected (also when updating uboot, but I doubt anyone really does that).

**Comment by mkresin** [comment]
Merged and backported to 18.06. Thanks!

And yes, the github syntax highlighting for patches patching patches is batshit crazy.
